lib/lrmd/lrmd_client.c

```diff
@@ -2343,6 +2343,121 @@ lrmd_api_delete(lrmd_t * lrmd)
+ *                       the main loop, and on success the metadata will be in
+ *                       its result argument's action_stdout)
+ * \param[in] user_data  User data to pass to callback
+ *
+ * \return Standard Pacemaker return code
+ * \note This function is not a lrmd_api_operations_t method because it does not
+ *       need an lrmd_t object and does not go through the executor, but
+ *       executes the agent directly.
+ */
+int
+lrmd__metadata_async(lrmd_rsc_info_t *rsc,
+                     void (*callback)(int pid,
+                                      const pcmk__action_result_t *result,
+                                      void *user_data),
+                     void *user_data)
+{
+    svc_action_t *action = NULL;
+    struct metadata_cb *metadata_cb = NULL;
+    pcmk__action_result_t result = PCMK__UNKNOWN_RESULT;
+
+    CRM_CHECK(callback != NULL, return EINVAL);
+
+    if ((rsc == NULL) || (rsc->standard == NULL) || (rsc->type == NULL)) {
+        pcmk__set_result(&result, PCMK_OCF_NOT_CONFIGURED, PCMK_EXEC_ERROR,
+                         "Invalid resource specification");
+        callback(0, &result, user_data);
+        pcmk__reset_result(&result);
+        return EINVAL;
+    }
+
+    if (strcmp(rsc->standard, PCMK_RESOURCE_CLASS_STONITH) == 0) {
+        return stonith__metadata_async(rsc->type,
+                                       CRMD_METADATA_CALL_TIMEOUT / 1000,
+                                       callback, user_data);
+    }
+
+    action = services__create_resource_action(rsc->type, rsc->standard,
+                                              rsc->provider, rsc->type,
+                                              CRMD_ACTION_METADATA, 0,
+                                              CRMD_METADATA_CALL_TIMEOUT, NULL,
+                                              0);
+    if (action == NULL) {
+        pcmk__set_result(&result, PCMK_OCF_UNKNOWN_ERROR, PCMK_EXEC_ERROR,
+                         "Out of memory");
+        callback(0, &result, user_data);
+        pcmk__reset_result(&result);
+        return ENOMEM;
+    }
+    if (action->rc != PCMK_OCF_UNKNOWN) {
+        pcmk__set_result(&result, action->rc, action->status,
+                         services__exit_reason(action));
+        callback(0, &result, user_data);
+        pcmk__reset_result(&result);
+        services_action_free(action);
+        return EINVAL;
+    }
+
+    action->cb_data = calloc(1, sizeof(struct metadata_cb));
+    if (action->cb_data == NULL) {
+        services_action_free(action);
+        pcmk__set_result(&result, PCMK_OCF_UNKNOWN_ERROR, PCMK_EXEC_ERROR,
+                         "Out of memory");
+        callback(0, &result, user_data);
+        pcmk__reset_result(&result);
+        return ENOMEM;
+    }
+
+    metadata_cb = (struct metadata_cb *) action->cb_data;
+    metadata_cb->callback = callback;
+    metadata_cb->user_data = user_data;
+    if (!services_action_async(action, metadata_complete)) {
```

Comment:
Got it, I'll take a closer look in a bit

Reply:
There are a couple of cases that can cause `services_action_async()` to return before calling `execute_action()`, though I don't think we're vulnerable to either of them here.

If `execute_action() -> services__execute_whatever()` doesn't return `pcmk_rc_ok`, then the caller is responsible for freeing the action. Currently we free `action` (via `services_action_free()`) but not `action->cb_data`.

If we do make it into the mainloop, then I think we're guaranteed to run the `metadata_complete()` callback as long as we don't hit any of the `CRM_CHECK` conditions along the way and/or have a bigger problem with the mainloop.

---

I think we'd want to free `metadata_cb` outright if `!services_action_async(action, metadata_complete)`, alongside the `services_action_free()` call.

`action->cb`'s members (`callback` and `user_data`) can be left alone since they're specified as `[in]` params, so `lrmd__metadata_async()`'s caller is responsible for freeing them if they're dynamically allocated.

Reply:
> There are a couple of cases that can cause `services_action_async()` to return before calling `execute_action()`, though I don't think we're vulnerable to either of them here.

This will never be recurring, so that avoids one of them.

It could potentially hit the is_op_blocked() block, but that's fine, the action will definitely be finalized at some future point and the callback will be called then.

> 
> However, if `execute_action() -> services__execute_whatever()` doesn't return `pcmk_rc_ok`, then the caller is responsible for freeing the action. Currently we free `action` (via `services_action_free()`) but not `action->cb_data`.

I believe that any path that doesn't return `pcmk_rc_ok` will call `services__finalize_async_op()`, which will call the callback.

> 
> If we do make it into the mainloop, then I think we're guaranteed to run the `metadata_complete()` callback as long as we don't hit any of the `CRM_CHECK` conditions along the way and/or have a bigger problem with the mainloop.
> 
> I think we'd want to free `metadata_cb`/`action->cb_data` outright if `!services_action_async(action, metadata_complete)`, alongside the `services_action_free()` call.
> 
> `action->cb`'s members (`callback` and `user_data`) can be left alone since they're specified as `[in]` params, so `lrmd__metadata_async()`'s caller is responsible for freeing them if they're dynamically allocated.

Reply:
> I believe that any path that doesn't return `pcmk_rc_ok` (for async ops) will call `services__finalize_async_op()`, which will call the callback.

Pretty sure you're right, not sure how I missed that. Although `services__finalize_async_op()` also calls `services_action_free()`, provided it's not a recurring or sync op, so the `services_action_free()` we have now may be redundant. What caught my eye was that it looked like we freed the action and some of its members but not its `cb_data`.

If you wanted to be extra paranoid, you could add a return-early-on-null to the beginning of the callback, and then run the callback before `services_action_free()` if `!services_action_async`. Otherwise, I think we're good now.

Reply:
> > I believe that any path that doesn't return `pcmk_rc_ok` (for async ops) will call `services__finalize_async_op()`, which will call the callback.
> 
> Pretty sure you're right, not sure how I missed that. Although `services__finalize_async_op()` also calls `services_action_free()`, provided it's not a recurring or sync op, so the `services_action_free()` we have now may be redundant.

That would be a double free, but `services__finalize_async_op()` returns `pcmk_rc_ok` if it frees the action, so we won't get there.

This code is some of the most confusing in Pacemaker and I hope we get a chance to redesign it at some point, but that's likely a long ways off.

Reply:
> That would be a double free

Ah yeah there aren't a lot of null checks in `services_action_free()`.

If somehow `services__finalize_async_op()` hasn't freed the action (shouldn't happen), and thus we get into the `if` block, then it *also* hasn't run the callback to free `action->cb_data`... unless it's a recurring op, which it won't be. Is that right? That's why I suggested that we should free both if we're going to free either inside that block

Reply:
Right, knowing that it's asynchronous and non-recurring eliminates a lot of paths. The libcrmservice code has been carefully designed so that the callback is called and the action is freed under specific conditions as described in the doxygen blocks. For non-recurring asynchronous ops, finalize always calls the callback, frees the action, and returns ok (which means services_action_async() returns true).

Reply:
But I see your point, if it were possible for services_action_async() to return false, then the callback wouldn't have been called. Maybe we should just CRM_CHECK() that the return value is true.

Reply:
Actually finalize *should* call the callback even on failure (excluding the invalid arguments case), so I think we're good

Reply:
Practically speaking yeah we're good. For a non-recurring, async op, there's no path through which finalize returns an error or doesn't call the callback... but there's also no path through which finalize doesn't call `services_action_free()` -- yet we call that in `lrmd__metadata_async()` even though it's unreachable. We should never encounter the case `!services_action_async` there.

At this point we're in agreement that we're safe from memory management issues here, so it's just a semantic question (i.e., should we even bother calling `services_action_free()` if we can't even reach that block, and should we also free the `cb_data` if we do).

I'm good with moving on haha